backend/app/notify/email.go

```diff
@@ -0,0 +1,380 @@
+	Quit() error
+	Close() error
+}
+
+type emailMessage struct {
+	message string
+	to      string
+}
+
+// msgTmplData store data for message from request template execution
+type msgTmplData struct {
+	From      string
+	To        string
+	Orig      string
+	Link      string
+	PostTitle string
+}
+
+// verifyTmplData store data for verification message template execution
+type verifyTmplData struct {
+	User  string
+	Email string
+	Token string
+	Site  string
+}
+
+const defaultEmailTimeout = 10 * time.Second
+const defaultFlushDuration = time.Second * 30
+const defaultEmailTemplate = `{{.From}}{{if .To}} → {{.To}}{{end}}
+
+{{.Orig}}
+
+↦ <a href="{{.Link}}">{{if .PostTitle}}{{.PostTitle}}{{else}}original comment{{end}}</a>
+`
+
+const defaultEmailVerificationTemplate = `Confirmation for {{.User}} {{.Email}}, site {{.Site}}
+
+Token: {{.Token}}
+`
+
+//NewEmail makes new Email object, returns it even in case of problems
+// (e.MsgTemplate parsing error or error while testing smtp connection by credentials provided in params)
+func NewEmail(params EmailParams) (*Email, error) {
+	var err error
+	res := Email{EmailParams: params}
+	if res.TimeOut <= 0 {
+		res.TimeOut = defaultEmailTimeout
+	}
+	if res.FlushDuration <= 0 {
+		res.FlushDuration = defaultFlushDuration
+	}
+	if res.BufferSize <= 0 {
+		res.BufferSize = 1
+	}
+	if res.MsgTemplate == "" {
+		res.MsgTemplate = defaultEmailTemplate
+	}
+	if res.VerificationTemplate == "" {
+		res.VerificationTemplate = defaultEmailVerificationTemplate
+	}
+	// unbuffered send channel for sending messages to autoFlush goroutine
+	res.submit = make(chan emailMessage)
+
+	log.Printf("[DEBUG] create new email notifier for server %s with user %s, timeout=%s",
+		res.Host, res.Username, res.TimeOut)
+
+	// initialise templates
+	res.msgTmpl, err = template.New("messageFromRequest").Parse(res.MsgTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse message template")
+	}
+	res.verifyTmpl, err = template.New("messageFromRequest").Parse(res.VerificationTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse verification template")
+	}
+
+	// establish test connection
+	testSMTPClient, err := res.client()
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't establish test connection")
+	}
+	if err = testSMTPClient.Quit(); err != nil {
+		log.Printf("[WARN] failed to send quit command to %s:%d, %v", res.Host, res.Port, err)
+	}
+	if err = testSMTPClient.Close(); err != nil {
+		log.Printf("[WARN] can't close smtp connection, %v", err)
+	}
+	return &res, err
+}
+
+// Send email from request to request.parentUserEmail if it's set, otherwise do nothing and return nil, thread safe
+// do not returns sending error, except following cases:
+// 1. (likely impossible) template execution error from email message creation from request
+// 2. message dropped without sending in case of closed ctx
+func (e *Email) Send(ctx context.Context, req request) error {
```

Comment:
I don't think this is a good idea to make the buffered email clinet to be so much aware about higher levels. To me it looks like all it suppose to do is to accept the set of arguments in terms of email operation, i.e. no request but emails, from, to and so on. You are mixing two levels here - low-level email sender (with auto flash and buffering) and some higher level email service, transating requests and dealing with the sender.

Reply:
note: I'm not saying this is really must-have separation. The code here is simple enought and may work as is, just my initial impression.

Reply:
Addressed by 03cd00694479537723f882eaa2c3356b11ab939c, is it what you had in mind?